### PR TITLE
Enhance mock fleet-server and add --proxy-url integration test 

### LINKED ISSUE #2482: Create an integration test for the fleet proxy set by --proxy-url and in the policy
# Issue #

Follow up task from https://github.com/elastic/elastic-agent/pull/2468

**Describe the enhancement:**

Create an integration test to ensure the proxy set by `--proxy-url` on install/enrol isn't overwritten by an empty proxy defined on the fleet policy. However a valid proxy set in the policy should take precedence over a proxy defined by the cli flag.

the test should roughly follow the steps below:

 - Setup a proxy between the agent and a mock fleet-server
 - do not define a proxy in the policy
 - install/enrol the agent using the --proxy-url flag
 - while no proxy is set in the policy, the agent should use the proxy provided by --proxy-url flag
 - set a different (and valid) proxy in the policy
 - check the agent is using the policy provided by the policy
 
Ideas:
 - manually create the proxies within the tests so we can easily see and assert what is going through the proxy 

**Describe a specific use case for the enhancement or feature:**

A proxy set by the cli flag `--proxy-url` is local to the agent, what allows agents in the same policy to use different proxies, what isn't possible when defining the proxy in the policy. The latter case, all agents must use the same proxy.

# Definition of done # 

- [ ] Integration test using our new framework is in place to validate the mentioned scenario.
- [ ] The test also uses SOCKS5 proxy https://github.com/elastic/elastic-agent/issues/2964

### PR DESCRIPTION
**Pull request by AndersonQ**
### Extends `testing/fleetservertest` to work out of the box with an elastic-agent and a fake component:

See [proxy_url_test.go](testing/integration/proxy_url_test.go) for examples on how to use it all the enhancements.

Use cases:

#### test simple agent <-> fleet-serve interactions: Done
A single fleet-server handler involved.
##### Minimal requirements:
 - [x] easy to implement fleet-server handler
 - [x] run it as a valid HTTP server
Nice  have: New Handler functions taking  a few options to tailor the handler for
each test. Check [`handlers.go`](handlers.go) for the available paths and handlers.

#### Run the agent against the mock fleet-server
Allow to e2e test the Agent without using a real fleet-server which allows:
 - test error cases / unhappy paths / hard to reproduce states. Eg. fleet-server sends an invalid policy
 - develop and test features before they're fully implemented on fleet-server

in order run the agent against the mock fleet-server, it needs to:
 - [x] have the states, enroll, checkin and ack handlers implemented and "working together":
   - [x]  it enrols one agent
   - [x]  the checkins:
     - [x]  are for the enrolled agent
     - [x]  might be set to fail
     - [x]  actions can be added and later sent to the agent
     - [x]  nice to have:
       - [x]  choose how long the checkin handler takes to send the repose (simulate long poll)
   - [x]  track acks:
     - [x]  once an action is acked, it won't be sent again
     - [x]  acks can be set to fail
 - ~~[ ]  allowed integrations (no need for real ES):~~
     - ~~[ ]  fake component~~
     - ~~[ ]  fake shipper~~
 - [x]  an API key and Enrollment token: both ignored, there is no authentication

### Add `proxytest.Proxy`, a naive proxy intended for tests which allows to configure URL rewrite and check all the requests made to it.
See [proxy_url_test.go](testing/integration/proxy_url_test.go) for examples on how to use it.

### Add E2E tests for fleet-server's proxy:

 - [x] `--proxy-url` set and no proxy in the policy
 - [x] `--proxy-url` set and empty proxy in the policy
 - [x] `--proxy-url` set and valid proxy in the policy
 - [x] no `--proxy-url` set and valid proxy in the policy
 - [x] proxy set, the remove proxy from policy

## Why is it important?

Enable us to test better agent->fleet-server interactions without the need to deploy the whole stack as well as to test unusual corner cases which are had to reproduce with a real fleet-server.

Also the tests for the agent's fleet-server proxy ~should~ will ensure we do not introduce any regression.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- [x] I have added an integration test or an E2E test

## How to test this PR locally

See [proxy_url_test.go](testing/integration/proxy_url_test.go) for examples on how to use it all the mock fleet-server enhancements and the `proxytest.Proxy`.

### To use the mock fleet server
 - Comment out the test skip of `TestRunFleetServer`
 - run the test: `go test -v -run TestRunFleetServer ./testing/fleetservertest`
 - use the shown server address as the `--url` enroll parameter and any value for `--enrollment-token`
 - the agent successfully enroll.

- run the server:
```
❯ go test -v -run TestRunFleetServer ./testing/fleetservertest
=== RUN   TestRunFleetServer
listening on: "http://[::]:42833"
press CTRL + C or kill the test to stop
```

- enroll the agent
```
root@elastic-agent-dev:~# sudo ./elastic-agent install --url=http://192.168.56.1:42833 --enrollment-token=anythingWIllDO -nif
```

### To run the integration tests:

 - remotely:
```
GOTEST_FLAGS="-test.run TestProxyURL" mage integration:test
```

 - locally: **be warned it should be done on a VM as it'll replace any installed agent**
```
vagrant up elastic-agent
vagrant ssh elastic-agent

cd /vagrant
TEST_DEFINE_PREFIX=gambiarra go test -run TestProxyURL -v -tags integration -timeout 0 ./testing/integration/
```

## Related issues

- Relates to #2482

## Logs

 - agent vagrant VM
```
root@elastic-agent-dev:~# sudo ./elastic-agent install --url=http://192.168.56.1:42833 --enrollment-token=anythingWIllDO -nif
sudo: ./elastic-agent: command not found
root@elastic-agent-dev:~# cd ./elastic-agent-8.10.0-SNAPSHOT-linux-x86_64/
root@elastic-agent-dev:~/elastic-agent-8.10.0-SNAPSHOT-linux-x86_64# sudo ./elastic-agent install --url=http://192.168.56.1:42833 --enrollment-token=anythingWIllDO -nif
Installing in non-interactive mode.{"log.level":"warn","@timestamp":"2023-07-03T17:35:27.908Z","log.logger":"tls","log.origin":{"file.name":"tlscommon/tls_config.go","file.line":104},"message":"SSL/TLS verifications disabled.","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-07-03T17:35:27.919Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":478},"message":"Starting enrollment to URL: http://192.168.56.1:42833/","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-07-03T17:35:29.366Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":276},"message":"Successfully triggered restart on running Elastic Agent.","ecs.version":"1.6.0"}
Successfully enrolled the Elastic Agent.
Elastic Agent has been successfully installed.
root@elastic-agent-dev:~/elastic-agent-8.10.0-SNAPSHOT-linux-x86_64# 
```

 - fleet server mock
```
❯ go test -v -run TestRunFleetServer ./testing/fleetservertest
=== RUN   TestRunFleetServer
listening on: "http://[::]:42833"
press CTRL + C or kill the test to stop
    server.go:127: 200 - POST /api/fleet/agents/enroll? HTTP/1.1 192.168.56.42:36772
    server.go:127: 200 - POST /api/fleet/agents/agentID/checkin? HTTP/1.1 192.168.56.42:49162
    server.go:127: 200 - POST /api/fleet/agents/agentID/checkin? HTTP/1.1 192.168.56.42:49162
    server.go:127: 200 - POST /api/fleet/agents/agentID/checkin? HTTP/1.1 192.168.56.42:49162
    server.go:127: 200 - POST /api/fleet/agents/agentID/checkin? HTTP/1.1 192.168.56.42:49162

```

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

**Comment by cmacknz** [comment]
> Right I'd need ES/output information but I cannot get it from the framework without changing it. I want to avoid starting changing things before we align better how to do it and also double check you we're on the same page

Do you need the information for the actual ES/output? Or is that mocked too? These credentials should be coming back from the `define.Require` statement at the start of the test.

Examples:

https://github.com/elastic/elastic-agent/blob/a5bfb663b851d7a7a3e9834eb3ba032d80d8c51e/testing/integration/upgrade_test.go#L40

The set of what it returns:

https://github.com/elastic/elastic-agent/blob/a5bfb663b851d7a7a3e9834eb3ba032d80d8c51e/pkg/testing/define/define.go#L41-L55

Using the kibana client:

https://github.com/elastic/elastic-agent/blob/a5bfb663b851d7a7a3e9834eb3ba032d80d8c51e/testing/integration/upgrade_test.go#L149

File: testing/fleetservertest/handlers.go [hunk 1]
```diff
@@ -61,43 +58,180 @@ func NewPathUploadComplete(agentID string) string {
...
+		// simulate long pool
+		time.Sleep(data.Delay)
+
 		return &resp, nil
 	}
 }
 
-func NewStatusHandlerHealth() func(ctx context.Context) (*StatusResponse, *HTTPError) {
-	return func(ctx context.Context) (*StatusResponse, *HTTPError) {
+func NewHandlerStatusHealth() func(ctx context.Context, _ *Handlers) (*StatusResponse, *HTTPError) {
```
Comment (ycombinator) [inline testing/fleetservertest/handlers.go:233]:
```suggestion
func NewHandlerStatusHealthy() func(ctx context.Context, _ *Handlers) (*StatusResponse, *HTTPError) {
```

File: testing/fleetservertest/handlers.go [hunk 2]
```diff
@@ -106,7 +240,7 @@ func NewStatusHandlerHealth() func(ctx context.Context) (*StatusResponse, *HTTPE
 	}
 }
 
-func NewStatusHandlerUnhealth() func(ctx context.Context) (*StatusResponse, *HTTPError) {
+func NewHandlerStatusUnhealth() func(ctx context.Context) (*StatusResponse, *HTTPError) {
```
Comment (ycombinator) [inline testing/fleetservertest/handlers.go:243]:
```suggestion
func NewHandlerStatusUnhealthy() func(ctx context.Context) (*StatusResponse, *HTTPError) {
```

File: testing/fleetservertest/server.go [hunk 3]
```diff
@@ -11,172 +11,123 @@ import (
...
-
-	return &UploadComplete200Response{Status: "ok"}, nil
-}
-
-// Handlers binds http requests to an api service and writes the service results to the http response
+// Handlers holds the handlers for the fleet-api, see https://petstore.swagger.io/?url=https://raw.githubusercontent.com/elastic/fleet-server/main/model/openapi.yml
+// for rendered OpenAPI definition. If any of the handlers are nil, a
+// http.StatusNotImplemented is returned for the route.
+//
+// Authentication is made extracting he API key or enrollment token,
```
Comment (ycombinator) [inline testing/fleetservertest/server.go:23]:
```suggestion
// Authentication is done by extracting the API key or enrollment token,
```

File: testing/fleetservertest/server.go [hunk 4]
```diff
@@ -11,172 +11,123 @@ import (
...
+// from the HeaderAuthorization header and compared against API.APIKey or
+// API.EnrollmentToken. API.EnrollmentToken is used for Enroll requests and
+// API.APIKey for all others.
```
Comment (ycombinator) [inline testing/fleetservertest/server.go:26]: ❤️ the clear documentation on authentication.

File: testing/fleetservertest/server.go [hunk 5]
```diff
@@ -11,172 +11,123 @@ import (
...
+// TODO(Anderson): fix me!
 type Handlers struct {
-	api API
+	// AgentID is the ID of agent communicating with this fleet-server:
+	//  - on Enrol this ID is set and returned to the enrolling agent,
```
Comment (ycombinator) [inline testing/fleetservertest/server.go:30]:
```suggestion
	//  - on Enroll this ID is set and returned to the enrolling agent,
```

File: testing/fleetservertest/server.go [hunk 6]
```diff
@@ -11,172 +11,123 @@ import (
...
+		id string,
+		userAgent string,
+		acceptEncoding string,
+		checkinRequest CheckinRequest) (*CheckinResponse, *HTTPError)
+
+	EnrollFn func(
+		ctx context.Context,
+		h *Handlers,
+		userAgent string,
+		enrolmentToken string,
```
Comment (ycombinator) [inline testing/fleetservertest/server.go:60]:
```suggestion
		enrollmentToken string,
```

File: testing/fleetservertest/server.go [hunk 7]
```diff
@@ -300,24 +268,32 @@ func (h *Handlers) AgentCheckin(w http.ResponseWriter, r *http.Request) {
...
+
 	userAgentParam := r.Header.Get("User-Agent")
 	enrollRequestParam := EnrollRequest{}
 	d := json.NewDecoder(r.Body)
 	if err := d.Decode(&enrollRequestParam); err != nil {
 		respondAsJSON(http.StatusBadRequest, HTTPError{
 			StatusCode: http.StatusBadRequest,
-			Error:      http.StatusText(http.StatusBadRequest),
-			Message:    fmt.Sprintf("%v", err),
+			Message:    fmt.Sprintf("could not decode enrol request: %v", err),
```
Comment (ycombinator) [inline testing/fleetservertest/server.go:284]:
```suggestion
			Message:    fmt.Sprintf("could not decode enroll request: %v", err),
```

File: testing/fleetservertest/fleetserver.go [hunk 8]
```diff
@@ -5,63 +5,114 @@
...
+	agentID string,
+	policyID string,
+	nextAction func() (CheckinAction, *HTTPError),
+	acker func(id string) (AckResponseItem, bool),
+	opts ...Option) *Server {
+
+	handlers := &Handlers{
+		APIKey:          apiKey.Key,
+		EnrollmentToken: enrolmentToken,
+		AgentID:         agentID, // as there is no enrol, the agentID needs to be manually set
```
Comment (ycombinator) [inline testing/fleetservertest/fleetserver.go:110]:
```suggestion
		AgentID:         agentID, // as there is no enroll, the agentID needs to be manually set
```

File: testing/fleetservertest/fleetserver_test.go [hunk 9]
```diff
@@ -40,28 +146,523 @@ func ExampleNewServer_status() {
...
+			},
+		},
 	})
+	if err != nil {
+		panic(fmt.Sprintf("could not execute enrol command: %v", err))
+	}
+
+	bs, err := json.MarshalIndent(resp, "", "  ")
+	if err != nil {
+		panic(fmt.Sprintf("could not marshal enrol response: %v", err))
```
Comment (ycombinator) [inline testing/fleetservertest/fleetserver_test.go:255]:
```suggestion
		panic(fmt.Sprintf("could not marshal enroll response: %v", err))
```

File: testing/fleetservertest/fleetserver.go [hunk 10]
```diff
@@ -5,63 +5,114 @@
...
 }
 
 // NewServer returns a new started *httptest.Server mocking the Fleet Server API.
-// If a route is called and its handler (the *Fn field) is nil a.
+// If a route is called and its handler (the *Fn field) is nil a
 // http.StatusNotImplemented error will be returned.
-func NewServer(api API) *httptest.Server {
-	mux := NewRouter(Handlers{api: api})
+func NewServer(h *Handlers, opts ...Option) *Server {
+	os := options{}
```
Comment (pchila) [inline testing/fleetservertest/fleetserver.go:33]: Nit: could you choose a differennt var name? For a moment in the code below I thought you were using the `os` package :sweat_smile:

File: testing/fleetservertest/fleetserver_test.go [hunk 11]
```diff
@@ -6,21 +6,127 @@ package fleetservertest
...
+		AgentID:  agentID,
+		ActionID: actionID,
+		PolicyID: policyID,
+		// FleetHosts needs to be changed after the server is running, so we can
+		// get the port the server is listening on. Therefore, the action generator
+		// captures the 'fleetHosts' variable, so it can read the real fleet-server
+		// address from it.
+		FleetHosts: `"host1", "host2"`,
+		SourceURI:  "http://source.uri",
+		CreatedAt:  "2023-05-31T11:37:50.607Z",
```
Comment (pchila) [inline testing/fleetservertest/fleetserver_test.go:60]: In the action policy template above the `created_at` field is not templated (maybe we need to modify it? )
I will add a suggestion for that that you can use if you want

File: testing/fleetservertest/checkin.go [hunk 12]
```diff
@@ -4,8 +4,126 @@
...
+	actionPolicyChangeTmpl = `
+    {
+      "agent_id": "{{.AgentID}}",
+      "created_at": "2023-05-31T11:37:50.607Z",
```
Comment (pchila) [inline testing/fleetservertest/checkin.go:62]:
```suggestion
      "created_at": "{{.CreatedAt}}",
```

File: testing/fleetservertest/fleetserver_test.go [hunk 13]
```diff
@@ -6,21 +6,127 @@ package fleetservertest
...
 	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
 )
 
+// TestRunFleetServer shows how to configure and run a fleet-server capable of
+// enrolling and send actions to a single agent.
+// For a detailed explanation of each piece and how to put them work together,
+// look at ExampleNewServer_checkin_and_ackWithAcker.
+// There is no authentication and the enrollment token is ignored, thus it can
+// be any value.
+func TestRunFleetServer(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (pchila) [reply testing/fleetservertest/fleetserver_test.go:28]: +1

File: testing/fleetservertest/fleetserver_test.go [hunk 14]
```diff
@@ -40,28 +146,523 @@ func ExampleNewServer_status() {
...
+	if err != nil {
+		panic(fmt.Sprintf("failed executing checkin: %v", err))
+	}
+	fmt.Printf("%#v\n", respAck)
+
+	// Output:
+	// &fleetapi.AckResponse{Action:"acks", Errors:false, Items:[]fleetapi.AckResponseItem{fleetapi.AckResponseItem{Status:200, Message:"OK"}, fleetapi.AckResponseItem{Status:200, Message:"OK"}}}
+}
+
+func ExampleNewServer_enrol() {
```
Comment (pchila) [inline testing/fleetservertest/fleetserver_test.go:217]: For consistency with the other suggestions

```suggestion
func ExampleNewServer_enroll() {
```

File: testing/fleetservertest/handlers.go [hunk 15]
```diff
@@ -61,43 +58,180 @@ func NewPathUploadComplete(agentID string) string {
...
 		if err != nil {
 			return nil, &HTTPError{
 				StatusCode: http.StatusInternalServerError,
-				Error:      err.Error(),
-				Message:    "failed to unmarshal policy",
+				Message:    fmt.Sprintf("failed to unmarshal policy: %v", err),
 			}
 		}
 
+		// simulate long pool
```
Comment (pchila) [inline testing/fleetservertest/handlers.go:226]: I wish it was a pool :ocean: :beach_umbrella:  :wink: 
```suggestion
		// simulate long poll
```

File: testing/fleetservertest/fleetserver_test.go [hunk 16]
```diff
@@ -6,21 +6,127 @@ package fleetservertest
...
+			APIKey: apiKey.String(),
+			Hosts:  `"https://my.clould.elstc.co:443"`,
+			Type:   "elasticsearch"},
+	}
+
+	nextAction := func() (CheckinAction, *HTTPError) {
+		defer func() { actionsIdx++ }()
+		tmpl.FleetHosts = fleetHosts
+
+		actions, err := NewActionPolicyChangeWithFakeComponent(tmpl)
```
Comment (ycombinator) [inline testing/fleetservertest/fleetserver_test.go:75]: I think this just returns a single action?

```suggestion
		action, err := NewActionPolicyChangeWithFakeComponent(tmpl)
```

File: testing/fleetservertest/fleetserver_test.go [hunk 17]
```diff
@@ -40,28 +146,523 @@ func ExampleNewServer_status() {
 
 func ExampleNewServer_checkin() {
 	agentID := "agentID"
+	tmpl := TmplPolicy{ActionID: "anActionID"}
+
+	actions, err := NewActionPolicyChangeWithFakeComponent(tmpl)
```
Comment (ycombinator) [inline testing/fleetservertest/fleetserver_test.go:151]: I think this just returns a single action?

```suggestion
	action, err := NewActionPolicyChangeWithFakeComponent(tmpl)
```

File: docs/test-framework-dev-guide.md [hunk 18]
```diff
@@ -49,6 +49,16 @@ pass `[testName]` to `go test` as `--run=[testName]`.
...
+### Manually running the tests
+
+If you want to run the tests manually, skipping the test runner, set the
+`TEST_DEFINE_PREFIX` environment variable to any value and run your tests normally
+with `go test`. E.g.:
+
+```shell
+TEST_DEFINE_PREFIX=gambiarra go test -v -tags integration -run TestProxyURL ./testing/integration/
+```
+
```
Comment (ycombinator) [inline docs/test-framework-dev-guide.md:61]: Thanks for adding this section.  

For completeness, you may want to also note that additional variables may need to be set, depending on the test's requirements.  For example, if the test is going to need an Elastic stack, `ELASTICSEARCH_HOST`, `ELASTICSEARCH_USERNAME`, `ELASTICSEARCH_PASSWORD`, `KIBANA_HOST`, `KIBANA_USERNAME`, and `KIBANA_PASSWORD` will also need to be set.

File: testing/fleetservertest/handlers.go [hunk 19]
```diff
@@ -61,43 +58,190 @@ func NewPathUploadComplete(agentID string) string {
...
+}
+
+// Acker is a function that for each actionID must return a AckResponseItem for
+// that action and if this ack errored or not.
+type Acker func(actionID string) (AckResponseItem, bool)
+
+// NewHandlerAckWithAcker takes an acker, a function that for each actionID must
+// return the expected AckResponseItem for that action and if this ack errored
+// or not.
+// TODO(Anderson): fix me
```
Comment (ycombinator) [inline testing/fleetservertest/handlers.go:90]: Does this `TODO` need more work in this PR? Or should this line be removed?

**Review by ycombinator — DISMISSED** [review]
Left a couple of minor suggestions, but overall LGTM!

File: docs/test-framework-dev-guide.md [hunk 20]
```diff
@@ -49,6 +49,20 @@ pass `[testName]` to `go test` as `--run=[testName]`.
...
+Tests with external dependencies might need more environment variables to be set
+when running they manually, such as `ELASTICSEARCH_HOST`, `ELASTICSEARCH_USERNAME`,
```
Comment (ycombinator) [inline docs/test-framework-dev-guide.md:63]: Typo.

```suggestion
when running them manually, such as `ELASTICSEARCH_HOST`, `ELASTICSEARCH_USERNAME`,
```

**Review by ycombinator — COMMENTED** [review]
Just a small typo fix, will approve after that!

**Review by ycombinator — APPROVED** [review]
LGTM - good stuff!
